toolchain/check/impl_validation.cpp

```diff
@@ -0,0 +1,456 @@
+    did_diagnose = diagnose_unmatchable_impl(impl_b, impl_a_id, impl_a);
+  }
+  if (impl_b.is_final && !did_diagnose) {
+    did_diagnose = diagnose_unmatchable_impl(impl_a, impl_b_id, impl_b);
+  }
+  if (did_diagnose) {
+    return true;
+  }
+
+  return false;
+}
+
+// Final impls that overlap in their type structure must be in the
+// same file.
+//
+// Returns true if an error was diagnosed.
+static auto DiagnoseFinalImplsOverlapInDifferentFiles(Context& context,
+                                                      const ImplInfo& impl_a,
+                                                      const ImplInfo& impl_b)
+    -> bool {
+  if (impl_a.ir_id != impl_b.ir_id) {
+    CARBON_DIAGNOSTIC(
+        FinalImplOverlapsDifferentFile, Error,
+        "`final impl` overlaps with `final impl` from another file");
+    CARBON_DIAGNOSTIC(FinalImplOverlapsDifferentFileNote, Note,
+                      "imported `final impl` here");
+    if (impl_a.is_local) {
+      auto builder = context.emitter().Build(impl_a.latest_decl_id,
+                                             FinalImplOverlapsDifferentFile);
+      builder.Note(impl_b.latest_decl_id, FinalImplOverlapsDifferentFileNote);
+      builder.Emit();
+    } else {
+      auto builder = context.emitter().Build(impl_b.latest_decl_id,
+                                             FinalImplOverlapsDifferentFile);
+      builder.Note(impl_a.latest_decl_id, FinalImplOverlapsDifferentFileNote);
+      builder.Emit();
+    }
+    return true;
+  }
+
+  return false;
+}
+
+// Two final impls in the same file can not overlap in their type
+// structure if they are not in the same match_first block.
+//
+// TODO: Support for match_first needed here when they exist in the
+// toolchain.
+//
+// Returns true if an error was diagnosed.
+static auto DiagnoseFinalImplsOverlapOutsideMatchFirst(Context& context,
+                                                       const ImplInfo& impl_a,
+                                                       const ImplInfo& impl_b)
+    -> bool {
+  if (impl_a.is_local && impl_b.is_local) {
+    CARBON_DIAGNOSTIC(FinalImplOverlapsSameFile, Error,
+                      "`final impl` overlaps with `final impl` from same file "
+                      "outside a `match_first` block");
+    auto builder = context.emitter().Build(impl_b.latest_decl_id,
+                                           FinalImplOverlapsSameFile);
+    CARBON_DIAGNOSTIC(FinalImplOverlapsSameFileNote, Note,
+                      "other `final impl` here");
+    builder.Note(impl_a.latest_decl_id, FinalImplOverlapsSameFileNote);
+    builder.Emit();
+    return true;
+  }
+
+  return false;
+}
+
+static auto ValidateImplsForInterface(
+    Context& context, llvm::ArrayRef<ImplAndInterface> impls_and_interface)
+    -> void {
+  // Range over `SemIR::ImplId` only. It'd be nice to make this the function
+  // parameter but we don't have a concept to express that outside the (banned)
+  // std::ranges.
```

Comment:
I feel like the detail in this comment was more distracting than helpful.
```suggestion
  // Range over `SemIR::ImplId` only. Caller has ensured all of these are
  // for the same interface.
```

Reply:
It was more meant to be a TODO, but rather than rewriting it as such, done.